### ISSUE TITLE
pure_funcs calls are not inlined

### ISSUE THREAD

**Issue by RReverser** [terser/terser#528]
**Bug report or Feature request?**

Feature request.

**Version (complete output of `terser -V` or specific git commit)**

4.3.4

**Complete CLI command or `minify()` options used**

```bash
npx terser temp.js -c pure_funcs=["Uint8Array"],passes=2
```

**`terser` input**

```js
(function(u) {
  console.log(u);
})([1, 2, 3]);

new Uint8Array([1, 2, 3]);

(function foo(u) {
  console.log(u);
})(new Uint8Array([1, 2, 3]));
```

**`terser` output or error**

```js
console.log([1,2,3]),function(u){console.log(u)}(new Uint8Array([1,2,3]));
```

**Expected result**

```js
console.log([1,2,3]),console.log(new Uint8Array([1,2,3]));
```

---

As you can see, `pure_funcs` correctly works for DCE, but for some reason it's not recognised as a constant that can be inlined in other functions.

**Comment by RReverser** [comment]
Note that using explicit annotations in this context doesn't seem to work either.

**Comment by kzc** [comment]
```
$ terser input.js -c unsafe,passes=2,pure_funcs=[Uint8Array]
console.log([1,2,3]),console.log(new Uint8Array([1,2,3]));
```

**Comment by RReverser** [comment]
Hmm weird. I''m pretty sure I tried `unsafe` too but maybe not this particular combination.

Any reason why this transformation is considered unsafe?

**Comment by RReverser** [comment]
Okay, I see what happened - I minimised example a bit too much for the issue report, and haven't tried with same combination of all possible options.

The issue seems to be somehow specific to async functions, and only when used with `await`, e.g.:

```js
(async function foo(e) {
  doSomething(e);
})(new Uint8Array([1, 2, 3]));

(async function foo(e) {
  await doSomething(e);
})(new Uint8Array([1, 2, 3]));

(async function foo(e) {
  await 42;
  doSomething(e);
})(new Uint8Array([1, 2, 3]));

(async function foo(e) {
  doSomething(e);
  await 42;
})(new Uint8Array([1, 2, 3]));
```

with

```bash
npx terser ../temp.js -c passes=10,unsafe,pure_funcs='["Uint8Array"]' -b
```

outputs

```js
!async function(e) {
    doSomething(new Uint8Array([ 1, 2, 3 ]));
}(), async function(e) {
    await doSomething(e);
}(new Uint8Array([ 1, 2, 3 ])), async function(e) {
    await 42, doSomething(e);
}(new Uint8Array([ 1, 2, 3 ])), async function(e) {
    doSomething(new Uint8Array([ 1, 2, 3 ])), await 42;
}();
```

That is, constant is correctly inlined only if used before an `await` boundary, although it should be equally safe to inline it even after or with `await` for pure functions.

Note that this works correctly for primitives passed to such functions.

**Comment by kzc** [comment]
This issue now has a great deal of overlap with #529 - inlining async functions.

`pure_funcs` is working correctly. The post above demonstrates an expected behavior of `collapse_vars`. Out of an abundance of caution this optimization stops propagating values forward when it encounters any control flow mechanism such as await. If you want to get your example to work as desired just remove this line:

https://github.com/terser/terser/blob/942a9ef73a36c2a9f260be739be36e19633e732f/lib/compress/index.js#L1352

I cannot attest to whether such a change is safe in general though. On numerous occasions we've found optimizations thought to be perfectly safe proven to not be the case. Nothing is obvious in ECMAScript. Unit tests only go so far. This project would benefit from a large real world ES2018 application with extensive use of async/await with a comprehensive test suite to minify and test against. Otherwise, the change can be made and the Internet can debug it for us. After all, the LICENSE does state: "THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER “AS IS” AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED."

**Comment by RReverser** [comment]
I agree there is some overlap, but I think they're rather different. Constant propagation should be safe pretty much everywhere (as long as we're sure it's indeed a constant / pure expression), but, upon further thinking, inlining `await` should actually be an opt-in unsafe transformation, as it can be observable from outside due to different number of microtasks. 

So I'd keep these two separate for now.

**Comment by kzc** [comment]
Objects or arrays such as the one created by `new Uint8Array()` are never considered to be constant despite being marked with `pure_funcs` or being pure annotated with a comment. They are only considered to be *side effect free* at the time of creation since they can subsequently be mutated. `reduce_vars` deals with constant propagation, whereas `collapse_vars` propagates non-constant variable payloads such as `new Uint8Array` forward so long as there are no intervening side effects introduced by AST nodes or other related variables for the payload in question. 

As far as `await` goes, we all know that when transpiled to ES5 with `regenerator` it becomes a complicated state machine that returns to the main loop. But in original non-transpiled form I'm not sure what assumptions can be made about a non-constant `await`ed payload in general.

Remove the line of code mentioned in my last post and test against your async codebase and report your findings. The broader problem is a lack of large async applications with comprehensive test suites to test against to empirically demonstrate that such optimizations are valid.

`await`ing constants or side-effect-free payloads are not very common. This proposed optimization like many others are not terribly significant in reducing overall bundle size. Whitespace removal and symbol mangling accounts for 95% of the size reduction for most minified code - not elaborate code transforms. Minify users value correctness above all, followed by minify speed, with bundle size being the third priority.

**Comment by RReverser** [comment]
> The broader problem is a lack of large async applications with comprehensive test suites to test against to empirically demonstrate that such optimizations are valid.

This is perhaps part of more general discussion / issue, but maybe terser could benefit from running against official test262 suite, like e.g. Babel does? For example, https://github.com/bterlson/test262-harness seems useful - some tests that rely on e.g. `Function::toString` will need to be excluded, but others should provide a comprehensible coverage.

**Comment by kzc** [comment]
Porting the test262 suite to terser would indeed be useful. But many of the optimizations in question (collapse_vars, reduce_vars, inline, side_effects, evaluate, unsafe, pure_funcs, et al) would not be thoroughly exercised by such small code tests.

`uglify-js` tests against https://browserbench.org/JetStream1.1, a large ES5 codebase, with each commit. Someone could port JetStream2 to terser (minus the Wasm bits), but even that appears to have minimal async/await test coverage: https://browserbench.org/JetStream/simple/file-system.js

Projects like Parcel or Rollup could be used since they make use of async/await and have comprehensive test suites - if they were changed to target ES2019. 

Any of the aforementioned test suites and others exercising advanced ES could be ported to https://github.com/terser/terser-functional-tests. Contributions welcome.

**Comment by kzc** [comment]
The more I think about it, unfortunately we can't have this proposed `await` optimization under `unsafe` either. Despite the option name, `unsafe` is a collection of safe optimizations and is fairly widely used.

We cannot assume that code in the wild being minified is well written. Such code should still produce the same result even when minified. A contrived illustrative example without using terser:

```
$ cat await1.js 
var g = [11];
async function foo() {
    var a = await [2];
    var b = a[0];
    g[0] += b;
    a[0] *= g[0];
    console.log(a, g);
    return a;
}
g[0] *= 10;
(async () => {
    console.log(await foo());
})();
console.log(g[0]); // value changes with await optimization
```
```
$ cat await1.js | node-v12.13.0 
110
[ 224 ] [ 112 ]
[ 224 ]
```
Now let's implement the proposed `await` optimization on constants:
```
$ cat await2.js 
var g = [11];
async function foo() {
    var a = [2];  //  <-----  var a = await [2];
    var b = a[0];
    g[0] += b;
    a[0] *= g[0];
    console.log(a, g);
    return a;
}
g[0] *= 10;
(async () => {
    console.log(await foo());
})();
console.log(g[0]); // value changes with await optimization
```
```
$ cat await2.js | node-v12.13.0 
[ 224 ] [ 112 ]
112
[ 224 ]
```
The resultant output is different in both order and value because the control flow has been altered by changing `await [2]` to `[2]`.

If we were to include such an optimization it'd have to be under `unsafe_await`, defaulting to `false`. Note: despite the prefix being the same, the `unsafe_xxx` compress optimizations are completely unrelated to the `unsafe` option and their settings are independent of it.

**Comment by RReverser** [comment]
> Now let's implement the proposed await optimization on constants:

Hmm, are you commenting on the wrong issue? That's not the proposed optimisation here - here I'm just asking for inlining of the actual value. Check out example in https://github.com/terser/terser/issues/528#issuecomment-562678296.

(In the other issue, yes, it's unsafe and externally observable to inline await calls.)

**Comment by kzc** [comment]
There are many overlapping optimizations being discussed in #529 and #528. It wasn't clear what was meant by inlining in this context since this issue morphed into a discussion about `await` and no expected result was provided.  I suppose the invalid example above eliding the `await` altogether is more related to #529. I will comment there to that effect.

This issue's title is misleading since the same behavior would be seen with a side-effect-free array literal without `pure_funcs`. This issue would be better described as "values not propagated into `await` expressions". I still advise putting such functionality under a new flag `unsafe_await` because I am still not convinced of its correctness in the general case.

**Comment by RReverser** [comment]
> This issue's title is misleading since the same behavior would be seen with a side-effect-free array literal without pure_funcs.

Hmm, probably. As I said, I was minimising the example as much as I could - might be that I didn't remove some pieces of code that are still unnecessary to fully reproduce the problem.

> "values not propagated into await expressions"

This is confusing too, because, as I said above, literals like `42` are propagated as expected. I guess it's more like "non-constant expressions..." or "object expressions are not propagated"?
